diesel/src/expression/assume_not_null.rs

```diff
@@ -0,0 +1,60 @@
+use crate::backend::Backend;
+use crate::expression::TypedExpressionType;
+use crate::expression::*;
+use crate::query_builder::*;
+use crate::query_source::joins::ToInnerJoin;
+use crate::result::QueryResult;
+use crate::sql_types::{DieselNumericOps, IntoNotNullable};
+
+#[derive(Debug, Copy, Clone, DieselNumericOps, ValidGrouping)]
+pub struct AssumeNotNull<T>(T);
+
+impl<T> AssumeNotNull<T> {
+    pub fn new(expr: T) -> Self {
+        AssumeNotNull(expr)
+    }
+}
+
+impl<T> Expression for AssumeNotNull<T>
+where
+    T: Expression,
+    T::SqlType: IntoNotNullable,
+    <T::SqlType as IntoNotNullable>::NotNullable: TypedExpressionType,
+{
+    type SqlType = <T::SqlType as IntoNotNullable>::NotNullable;
+}
+
+impl<T, DB> QueryFragment<DB> for AssumeNotNull<T>
+where
+    DB: Backend,
+    T: QueryFragment<DB>,
+{
+    fn walk_ast(&self, pass: AstPass<DB>) -> QueryResult<()> {
+        self.0.walk_ast(pass)
+    }
+}
+
+/// AssumeNotNull can be used in where clauses everywhere, but can only be used in
+/// select clauses for outer joins.
```

Comment:
Is there any reason for this restriction? I mean it's totally fine and valid to have something like this:
```rust
some_table::table.filter(some_table::some_nullable_column.is_not_null()).select((some_table::some_nullable_column.assume_not_null()).load(…)?
```

Reply:
That's a mistake, it comes from the copy-pasting of `Nullable`.
I don't understand what it means in the context of `Nullable` though.

Reply:
I believe the original comment meant to express that you only can use `.select(some::column.nullable())`  on a query that actually does a left join. I'm not sure if that statement remains true after the `SqlType` rewrite.

Reply:
I don't think that is actually true anymore.